--- a/misc-utils/whereis.c
+++ b/misc-utils/whereis.c
@@ -66,59 +66,22 @@ static char *bindirs[] = {
    "/usr/etc",
    "/lib",
    "/usr/lib",
-   "/lib64",
-   "/usr/lib64",
-   "/usr/games",
-   "/usr/games/bin",
-   "/usr/games/lib",
-   "/usr/emacs/etc",
-   "/usr/lib/emacs/*/etc",
-   "/usr/TeX/bin",
-   "/usr/tex/bin",
-   "/usr/interviews/bin/LINUX",
-
-   "/usr/X11R6/bin",
-   "/usr/X386/bin",
-   "/usr/bin/X11",
-   "/usr/X11/bin",
-   "/usr/X11R5/bin",
-
-   "/usr/local/bin",
-   "/usr/local/sbin",
-   "/usr/local/etc",
-   "/usr/local/lib",
-   "/usr/local/games",
-   "/usr/local/games/bin",
-   "/usr/local/emacs/etc",
-   "/usr/local/TeX/bin",
-   "/usr/local/tex/bin",
-   "/usr/local/bin/X11",
-
-   "/usr/contrib",
-   "/usr/hosts",
-   "/usr/include",
-
-   "/usr/g++-include",
-
-   "/usr/ucb",
-   "/usr/old",
-   "/usr/new",
-   "/usr/local",
-   "/usr/libexec",
-   "/usr/share",
-
-   "/opt/*/bin",
-
+   "/media/DISK_A1/system/bin",
+   "/media/DISK_A1/system/sbin",
+   "/media/DISK_A1/system/usr/bin",
+   "/media/DISK_A1/system/usr/sbin",
+   "/media/DISK_A1/system/usr/libexec",
+   "/media/DISK_A1/system/lib",
+   "/media/DISK_A1/system/usr/lib",
+   "/media/DISK_A1/system/etc",
+   "/media/DISK_A1/system/etc/init.d",
+   "/media/DISK_A1/system/etc/firewall.d",
 	0
 };
 
 static char *mandirs[] = {
-	"/usr/man/*",
-	"/usr/share/man/*",
-	"/usr/X386/man/*",
-	"/usr/X11/man/*",
-	"/usr/TeX/man/*",
-	"/usr/interviews/man/mann",
+	"/media/DISK_A1/system/usr/man/*",
+	"/media/DISK_A1/system/usr/share/man/*",
 	0
 };
 
@@ -126,8 +89,6 @@ static char *srcdirs[]  = {
 	"/usr/src/*",
 	"/usr/src/lib/libc/*",
 	"/usr/src/lib/libc/net/*",
-	"/usr/src/ucb/pascal",
-	"/usr/src/ucb/pascal/utilities",
 	"/usr/src/undoc",
 	0
 };
