--- a/ext/standard/php_crypt_r.c
+++ b/ext/standard/php_crypt_r.c
@@ -42,7 +42,7 @@
 # include <Wincrypt.h>
 #endif
 
-#ifdef HAVE_ATOMIC_H /* Solaris 10 defines atomic API within */
+#if defined(HAVE_ATOMIC_H) && !defined(__UCLIBC__) /* Solaris 10 defines atomic API within */
 # include <atomic.h>
 #else
 # include <signal.h>
@@ -81,7 +81,7 @@ void _crypt_extended_init_r(void)
 {
 #ifdef PHP_WIN32
 	LONG volatile initialized = 0;
-#elif defined(HAVE_ATOMIC_H) /* Solaris 10 defines atomic API within */
+#elif defined(HAVE_ATOMIC_H) && !defined(__UCLIBC__) /* Solaris 10 defines atomic API within */
 	volatile unsigned int initialized = 0;
 #else
 	static volatile sig_atomic_t initialized = 0;
@@ -96,7 +96,7 @@ void _crypt_extended_init_r(void)
 		InterlockedIncrement(&initialized);
 #elif (defined(__GNUC__) && (__GNUC__ >= 4 && __GNUC_MINOR__ >= 2))
 		__sync_fetch_and_add(&initialized, 1);
-#elif defined(HAVE_ATOMIC_H) /* Solaris 10 defines atomic API within */
+#elif defined(HAVE_ATOMIC_H) && !defined(__UCLIBC__) /* Solaris 10 defines atomic API within */
 		membar_producer();
 		atomic_add_int(&initialized, 1);
 #endif
