diff -Nru icecast-2.3.3.orig/configure icecast-2.3.3/configure
--- icecast-2.3.3.orig/configure	2012-06-11 21:03:18.000000000 +0400
+++ icecast-2.3.3/configure	2012-12-11 11:00:55.000000000 +0400
@@ -12503,7 +12503,7 @@
 else
    cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 int
 main ()
 {
diff -Nru icecast-2.3.3.orig/m4/ogg.m4 icecast-2.3.3/m4/ogg.m4
--- icecast-2.3.3.orig/m4/ogg.m4	2012-06-11 20:45:25.000000000 +0400
+++ icecast-2.3.3/m4/ogg.m4	2012-12-11 11:02:41.000000000 +0400
@@ -29,7 +29,7 @@
       ])
 AC_CACHE_CHECK([for libogg], xt_cv_lib_ogg,
 [dnl
-OGG_LIBS="-logg"
+OGG_LIBS="-lvorbisidec"
 
 #
 # check if the installed Ogg is sufficiently new.
@@ -42,7 +42,7 @@
 LDFLAGS="$LDFLAGS $OGG_LDFLAGS"
 AC_TRY_LINK_FUNC(ogg_sync_init,
     [ xt_cv_lib_ogg=ok ],
-    [ AC_TRY_LINK([#include <ogg/ogg.h>],, 
+    [ AC_TRY_LINK([#include <tremor/ogg.h>],, 
         [ xt_cv_lib_ogg="pre v1.0, needs updating" ],
         [ xt_cv_lib_ogg="not found" ])
     ])
diff -Nru icecast-2.3.3.orig/m4/vorbis.m4 icecast-2.3.3/m4/vorbis.m4
--- icecast-2.3.3.orig/m4/vorbis.m4	2012-06-11 20:45:25.000000000 +0400
+++ icecast-2.3.3/m4/vorbis.m4	2012-12-11 11:17:57.000000000 +0400
@@ -38,9 +38,9 @@
             ])
 fi
 
-VORBIS_LIBS="-lvorbis"
-VORBISFILE_LIBS="-lvorbisfile"
-VORBISENC_LIBS="-lvorbisenc"
+VORBIS_LIBS="-lvorbisidec"
+VORBISFILE_LIBS="-lvorbisidec"
+VORBISENC_LIBS="-lvorbisidec"
 
 xt_save_LIBS="$LIBS"
 xt_save_LDFLAGS="$LDFLAGS"
@@ -58,18 +58,6 @@
             )
         ])
 
-if test "x$xt_lib_vorbis" = "xok"; then
-#
-# Now check if the installed Vorbis is sufficiently new.
-#
-AC_COMPILE_IFELSE([AC_LANG_PROGRAM([
-#include <vorbis/codec.h>
-#include <vorbis/vorbisenc.h>
-        ], [
-struct ovectl_ratemanage_arg a;
-])],,[xt_lib_vorbis="old version found"])
-AC_MSG_RESULT([$xt_lib_vorbis])
-fi
 CPPFLAGS="$xt_save_CPPFLAGS"
 LIBS="$xt_save_LIBS"
 LDFLAGS="$xt_save_LDFLAGS"
diff -Nru icecast-2.3.3.orig/src/format_flac.c icecast-2.3.3/src/format_flac.c
--- icecast-2.3.3.orig/src/format_flac.c	2012-06-11 20:45:20.000000000 +0400
+++ icecast-2.3.3/src/format_flac.c	2012-12-11 10:59:27.000000000 +0400
@@ -18,7 +18,7 @@
 #endif
 
 #include <stdlib.h>
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 #include <string.h>
 
 typedef struct source_tag source_t;
diff -Nru icecast-2.3.3.orig/src/format_kate.c icecast-2.3.3/src/format_kate.c
--- icecast-2.3.3.orig/src/format_kate.c	2012-06-11 20:45:20.000000000 +0400
+++ icecast-2.3.3/src/format_kate.c	2012-12-11 10:58:40.000000000 +0400
@@ -19,7 +19,7 @@
 
 #include <stdlib.h>
 #include <string.h>
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 #ifdef HAVE_KATE
 #include <kate/oggkate.h>
 #endif
diff -Nru icecast-2.3.3.orig/src/format_midi.c icecast-2.3.3/src/format_midi.c
--- icecast-2.3.3.orig/src/format_midi.c	2012-06-11 20:45:20.000000000 +0400
+++ icecast-2.3.3/src/format_midi.c	2012-12-11 10:59:43.000000000 +0400
@@ -18,7 +18,7 @@
 #endif
 
 #include <stdlib.h>
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 #include <string.h>
 
 typedef struct source_tag source_t;
diff -Nru icecast-2.3.3.orig/src/format_ogg.c icecast-2.3.3/src/format_ogg.c
--- icecast-2.3.3.orig/src/format_ogg.c	2012-06-11 20:45:20.000000000 +0400
+++ icecast-2.3.3/src/format_ogg.c	2012-12-11 10:58:30.000000000 +0400
@@ -24,7 +24,7 @@
 #include <stdlib.h>
 #include <string.h>
 
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 
 #include "refbuf.h"
 #include "source.h"
diff -Nru icecast-2.3.3.orig/src/format_ogg.h icecast-2.3.3/src/format_ogg.h
--- icecast-2.3.3.orig/src/format_ogg.h	2012-06-11 20:45:20.000000000 +0400
+++ icecast-2.3.3/src/format_ogg.h	2012-12-11 11:00:02.000000000 +0400
@@ -18,7 +18,7 @@
 #ifndef __FORMAT_OGG_H__
 #define __FORMAT_OGG_H__
 
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 #include "refbuf.h"
 #include "format.h"
 
diff -Nru icecast-2.3.3.orig/src/format_skeleton.c icecast-2.3.3/src/format_skeleton.c
--- icecast-2.3.3.orig/src/format_skeleton.c	2012-06-11 20:45:20.000000000 +0400
+++ icecast-2.3.3/src/format_skeleton.c	2012-12-11 10:59:17.000000000 +0400
@@ -19,7 +19,7 @@
 
 #include <stdlib.h>
 #include <string.h>
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 
 typedef struct source_tag source_t;
 
diff -Nru icecast-2.3.3.orig/src/format_speex.c icecast-2.3.3/src/format_speex.c
--- icecast-2.3.3.orig/src/format_speex.c	2012-06-11 20:45:20.000000000 +0400
+++ icecast-2.3.3/src/format_speex.c	2012-12-11 10:59:02.000000000 +0400
@@ -18,7 +18,7 @@
 #endif
 
 #include <stdlib.h>
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 #include <speex/speex_header.h>
 
 typedef struct source_tag source_t;
diff -Nru icecast-2.3.3.orig/src/format_theora.c icecast-2.3.3/src/format_theora.c
--- icecast-2.3.3.orig/src/format_theora.c	2012-06-11 20:45:20.000000000 +0400
+++ icecast-2.3.3/src/format_theora.c	2012-12-11 10:58:12.000000000 +0400
@@ -18,7 +18,7 @@
 #endif
 
 #include <stdlib.h>
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 #include <theora/theora.h>
 
 typedef struct source_tag source_t;
diff -Nru icecast-2.3.3.orig/src/format_vorbis.c icecast-2.3.3/src/format_vorbis.c
--- icecast-2.3.3.orig/src/format_vorbis.c	2012-06-11 20:45:20.000000000 +0400
+++ icecast-2.3.3/src/format_vorbis.c	2012-12-11 10:58:02.000000000 +0400
@@ -18,7 +18,7 @@
 #endif
 
 #include <stdlib.h>
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 #include <vorbis/codec.h>
 #include <memory.h>
 #include <string.h>
diff -Nru icecast-2.3.3.orig/src/source.c icecast-2.3.3/src/source.c
--- icecast-2.3.3.orig/src/source.c	2012-06-11 20:45:20.000000000 +0400
+++ icecast-2.3.3/src/source.c	2012-12-11 10:58:49.000000000 +0400
@@ -19,7 +19,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <sys/types.h>
-#include <ogg/ogg.h>
+#include <tremor/ogg.h>
 #include <errno.h>
 
 #ifndef _WIN32
