--- a/src/lftp.cc
+++ b/src/lftp.cc
@@ -510,7 +510,7 @@ int   main(int argc,char **argv)
    top_exec->SetStatusLine(new StatusLine(1));
    Log::global->SetCB(tty_clear);
 
-   source_if_exist(SYSCONFDIR"/lftp.conf");
+   source_if_exist("/media/DISK_A1/system/etc/lftp.conf");
    const char *home=getenv("HOME");
    if(home)
       source_if_exist(dir_file(home,".lftprc"));
--- a/src/lftp_ssl.cc
+++ b/src/lftp_ssl.cc
@@ -198,13 +198,13 @@ static const char *lftp_ssl_find_ca_file
 {
    // a few possible locations of ca-bundle.crt
    static const char *const ca_file_location[]={
-      "/etc/pki/tls/certs/ca-bundle.crt",
-      "/etc/certs/ca-bundle.crt",
-      "/usr/share/ssl/certs/ca-bundle.crt",
-      "/etc/ssl/certs/ca-certificates.crt",
-      "/usr/local/ssl/certs/ca-bundle.crt",
-      "/etc/apache/ssl.crt/ca-bundle.crt",
-      "/usr/share/curl/curl-ca-bundle.crt",
+      "/media/DISK_A1/system/etc/pki/tls/certs/ca-bundle.crt",
+      "/media/DISK_A1/system/etc/certs/ca-bundle.crt",
+      "/media/DISK_A1/system/usr/share/ssl/certs/ca-bundle.crt",
+      "/media/DISK_A1/system/etc/ssl/certs/ca-certificates.crt",
+      "/media/DISK_A1/system/usr/local/ssl/certs/ca-bundle.crt",
+      "/media/DISK_A1/system/etc/apache/ssl.crt/ca-bundle.crt",
+      "/media/DISK_A1/system/usr/share/curl/curl-ca-bundle.crt",
       0};
    for(int i=0; ca_file_location[i]; i++)
    {
