diff -U 3 -dHBbrN -- ppp-2.4.4.orig/pppd/plugins/rp-pppoe/plugin.c ppp-2.4.4/pppd/plugins/rp-pppoe/plugin.c
--- ppp-2.4.4.orig/pppd/plugins/rp-pppoe/plugin.c	2011-01-27 22:09:49.000000000 +0300
+++ ppp-2.4.4/pppd/plugins/rp-pppoe/plugin.c	2011-01-27 22:36:19.000000000 +0300
@@ -216,6 +216,7 @@
     }
     close(conn->sessionSocket);
     /* don't send PADT?? */
+    sendPADT(conn, NULL);
     close(conn->discoverySocket);
 }
 
