diff -U 3 -H -w -B -E -d -r -N -- uClibc-0.9.30.3.orig/libc/inet/resolv.c uClibc-0.9.30.3/libc/inet/resolv.c
--- uClibc-0.9.30.3.orig/libc/inet/resolv.c	2010-08-27 02:33:51.000000000 +0400
+++ uClibc-0.9.30.3/libc/inet/resolv.c	2010-08-27 02:39:24.000000000 +0400
@@ -225,7 +225,7 @@
 #define MAX_RECURSE 5
 #define REPLY_TIMEOUT 10
 #define MAX_RETRIES 3
-#define MAX_SERVERS 3
+#define MAX_SERVERS 8
 #define MAX_SEARCH 4
 
 #define MAX_ALIASES	5
@@ -737,80 +737,54 @@
  * is pretty lame.  I think these two variables can probably be de-global-ized,
  * which should eliminate the need for doing locking here...  Needs a closer
  * look anyways. */
-static int static_ns = 0;
-/* uint16: minimizing rw data size, even if code grows a tiny bit.
- * rw data costs more. */
-static uint16_t static_id = 1;
+static int id = 1;
 
-/* On entry:
- *  a.buf(len) = auxiliary buffer for IP addresses after first one
- *  a.add_count = how many additional addresses are there already
- *  outpacket = where to save ptr to raw packet? can be NULL
- * On exit:
- *  ret < 0: error, all other data is not valid
- *  a.add_count & a.buf: updated
- *  a.rdlength: length of addresses (4 bytes for IPv4)
- *  *outpacket: updated (packet is malloced, you need to free it)
- *  a.rdata: points into *outpacket to 1st IP addr
- *      NB: don't pass outpacket == NULL if you need to use a.rdata!
- *  a.atype: type of query?
- *  a.dotted: which name we _actually_ used. May contain search domains
- *      appended. (why the filed is called "dotted" I have no idea)
- *      This is a malloced string. May be NULL because strdup failed.
- */
-int attribute_hidden __dns_lookup(const char *name, int type,
-			int nscount, char **nsip,
-			unsigned char **outpacket,
-			struct resolv_answer *a)
+int attribute_hidden __dns_lookup(const char *name, int type, int nscount, char **nsip,
+			   unsigned char **outpacket, struct resolv_answer *a)
 {
-	int i, j, len, fd, pos, rc;
-#ifdef USE_SELECT
+	int i, j, len, pos, rc;
 	struct timeval tv;
-	fd_set fds;
-#else
-	struct pollfd fds;
-#endif
+	fd_set fdsets;
 	struct resolv_header h;
 	struct resolv_question q;
 	struct resolv_answer ma;
-	bool first_answer = 1;
-	unsigned retries = 0;
+	bool first_answer;
 	unsigned char *packet = malloc(PACKETSZ);
-	char *dns, *lookup = malloc(MAXDNAME);
-	int variant = -1;  /* search domain to append, -1 - none */
-	int local_ns = -1, local_id = -1;
-	bool ends_with_dot;
-	union {
-		struct sockaddr sa;
-#ifdef __UCLIBC_HAS_IPV4__
-		struct sockaddr_in sa4;
-#endif
+	char *lookup = malloc(MAXDNAME);
+	int variant;
+	struct sockaddr_in sa;
+	int fds[MAX_SERVERS], ifds, ipos, ipass, itrys, imaxfd;
+	int sdomains;
+	int local_id;
 #ifdef __UCLIBC_HAS_IPV6__
+	bool v6;
 		struct sockaddr_in6 sa6;
 #endif
-	} sa;
 
-	fd = -1;
+	ifds = 0;
+	for( ipos = 0; ipos < MAX_SERVERS; ipos++ ) fds[ipos] = -1;
 
-	if (!packet || !lookup || !nscount || !name[0])
-		goto fail;
+	if( !packet || !lookup || !nscount || !name[0] )  goto fail;
 
 	DPRINTF("Looking up type %d answer for '%s'\n", type, name);
 
-	ends_with_dot = (name[strlen(name) - 1] == '.');
+	itrys = MAX_RETRIES;
+	ifds = 0;
+	variant = -1;
+	while( itrys > 0 ) {
+		for( ipos = 0; ipos < ifds; ipos++ )
+			if( fds[ipos] != -1 ) {
+				close(fds[ipos]);
+				fds[ipos] = -1;
+			}
 
 	/* Mess with globals while under lock */
 	__UCLIBC_MUTEX_LOCK(mylock);
-	local_ns = static_ns % nscount;
-	local_id = static_id;
+	local_id = id;
+		ifds = MIN(MAX_SERVERS, nscount);
 	__UCLIBC_MUTEX_UNLOCK(mylock);
 
-	while (retries < MAX_RETRIES) {
-		if (fd != -1)
-			close(fd);
-
 		memset(packet, 0, PACKETSZ);
-
 		memset(&h, 0, sizeof(h));
 
 		++local_id;
@@ -821,119 +795,111 @@
 
 		DPRINTF("encoding header\n", h.rd);
 
-		i = __encode_header(&h, packet, PACKETSZ);
-		if (i < 0)
-			goto fail;
+		if( (i = __encode_header(&h, packet, PACKETSZ)) < 0 ) goto fail;
+
+		strncpy(lookup, name, MAXDNAME - 1);
+		lookup[MAXDNAME - 1] = 0;
 
-		strncpy(lookup, name, MAXDNAME);
-		__UCLIBC_MUTEX_LOCK(__resolv_lock);
-		/* nsip is really __nameserver[] which is a global that
-		   needs to hold __resolv_lock before access!! */
-		dns = nsip[local_ns];
-/* TODO: all future accesses to 'dns' were guarded by __resolv_lock too.
- * Why? We already fetched nsip[local_ns] here,
- * future changes to nsip[] by other threads cannot affect us.
- * We can use 'dns' without locking. If I'm wrong,
- * please explain in comments why locking is needed.
- * One thing that worries me is - what if __close_nameservers() free()s
- * dns under us? __resolv_lock'ing around accesses to dns won't help either,
- * as free() might occur between accesses!
- */
 		if (variant >= 0) {
+			__UCLIBC_MUTEX_LOCK(__resolv_lock);
 			if (variant < __searchdomains) {
-				strncat(lookup, ".", MAXDNAME);
-				strncat(lookup, __searchdomain[variant], MAXDNAME);
-			}
+				strncat(lookup, ".", MAXDNAME - 1);
+				strncat(lookup, __searchdomain[variant], MAXDNAME - 1);
+				lookup[MAXDNAME - 1] = 0;
 		}
 		__UCLIBC_MUTEX_UNLOCK(__resolv_lock);
+		}
 
 		DPRINTF("lookup name: %s\n", lookup);
-		q.dotted = lookup;
+		q.dotted = (char *)lookup;
 		q.qtype = type;
 		q.qclass = C_IN; /* CLASS_IN */
 
-		j = __encode_question(&q, packet+i, PACKETSZ-i);
-		if (j < 0)
-			goto fail;
-
+		if( (j = __encode_question(&q, packet + i, PACKETSZ - i)) < 0 ) goto fail;
 		len = i + j;
 
-		DPRINTF("On try %d, sending query to port %d of machine %s\n",
-				retries+1, NAMESERVER_PORT, dns);
-
-		sa.sa.sa_family = AF_INET;
-#ifdef __UCLIBC_HAS_IPV6__
-		//__UCLIBC_MUTEX_LOCK(__resolv_lock);
-		///* 'dns' is really __nameserver[] which is a global that
-		//   needs to hold __resolv_lock before access!! */
-		if (inet_pton(AF_INET6, dns, &sa.sa6.sin6_addr) > 0)
-			sa.sa.sa_family = AF_INET6;
-		//__UCLIBC_MUTEX_UNLOCK(__resolv_lock);
-#endif
+		ipass = 0;
 		/* Connect to the UDP socket so that asyncronous errors are returned */
+
+		for( ipos = 0; ipos < ifds; ipos++ ) {
 #ifdef __UCLIBC_HAS_IPV6__
-		if (sa.sa.sa_family == AF_INET6) {
-			sa.sa6.sin6_port = htons(NAMESERVER_PORT);
+		__UCLIBC_MUTEX_LOCK(__resolv_lock);
+		/* 'dns' is really __nameserver[] which is a global that
+		   needs to hold __resolv_lock before access!! */
+			v6 = inet_pton(AF_INET6, nsip[ipos], &sa6.sin6_addr) > 0 ? 1 : 0;
+		__UCLIBC_MUTEX_UNLOCK(__resolv_lock);
+
+		if (v6) {
+				if( (fds[ipos] = socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP)) != -1 ) {
+		    sa6.sin6_family = AF_INET6;
+		    sa6.sin6_port = htons(NAMESERVER_PORT);
 			/* sa6.sin6_addr is already here */
-		} else
-#endif
-#ifdef __UCLIBC_HAS_IPV4__
-		{
-			sa.sa4.sin_port = htons(NAMESERVER_PORT);
-			//__UCLIBC_MUTEX_LOCK(__resolv_lock);
-			///* 'dns' is really __nameserver[] which is a global that
-			//   needs to hold __resolv_lock before access!! */
-			sa.sa4.sin_addr.s_addr = inet_addr(dns);
-			//__UCLIBC_MUTEX_UNLOCK(__resolv_lock);
+					if( (rc = connect(fds[ipos], (struct sockaddr *) &sa6, sizeof(sa6))) < 0 ) {
+						close(fds[ipos]);
+						fds[ipos] = -1;
+					} else ipass = 1;
 		}
+		} else {
 #endif
-		fd = socket(sa.sa.sa_family, SOCK_DGRAM, IPPROTO_UDP);
-		if (fd < 0) {
-			retries++;
-			continue;
+				if( (fds[ipos] = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) != -1 ) {
+		    sa.sin_family = AF_INET;
+		    sa.sin_port = htons(NAMESERVER_PORT);
+		    __UCLIBC_MUTEX_LOCK(__resolv_lock);
+		    /* 'dns' is really __nameserver[] which is a global that
+		       needs to hold __resolv_lock before access!! */
+					sa.sin_addr.s_addr = inet_addr(nsip[ipos]);
+		    __UCLIBC_MUTEX_UNLOCK(__resolv_lock);
+				
+					if( (rc = connect(fds[ipos], (struct sockaddr *) &sa, sizeof(sa))) < 0 ) {
+						close(fds[ipos]);
+						fds[ipos] = -1;
+					} else ipass = 1; 
 		}
-		rc = connect(fd, &sa.sa, sizeof(sa));
-		if (rc < 0) {
-			if (errno == ENETUNREACH) {
-				/* routing error, presume not transient */
-				goto tryall;
+#ifdef __UCLIBC_HAS_IPV6__
 			}
-			/* retry */
-			retries++;
+#endif
+		}
+		
+		if( !ipass ) {
+			--itrys;
 			continue;
 		}
 
 		DPRINTF("Transmitting packet of length %d, id=%d, qr=%d\n",
 				len, h.id, h.qr);
 
-		send(fd, packet, len, 0);
+		FD_ZERO(&fdsets);
+		
+		imaxfd = 0;
+		for( ipos = 0; ipos < ifds; ipos++ ) {
+			if( fds[ipos] != -1 ) {
+				send(fds[ipos], packet, len, 0);
+				FD_SET(fds[ipos], &fdsets);
+				if( fds[ipos] > imaxfd ) imaxfd = fds[ipos];
+			}
+		}
 
-#ifdef USE_SELECT
-		FD_ZERO(&fds);
-		FD_SET(fd, &fds);
 		tv.tv_sec = REPLY_TIMEOUT;
 		tv.tv_usec = 0;
-		if (select(fd + 1, &fds, NULL, NULL, &tv) <= 0) {
+
+		if( select(imaxfd + 1, &fdsets, NULL, NULL, &tv) <= 0 ) {
 			DPRINTF("Timeout\n");
-			/* timed out, so retry send and receive,
-			 * to next nameserver on queue */
-			goto tryall;
+		    --itrys;
+		    continue;
 		}
-#else
-		fds.fd = fd;
-		fds.events = POLLIN;
-		if (poll(&fds, 1, REPLY_TIMEOUT * 1000) <= 0) {
-			DPRINTF("Timeout\n");
-			/* timed out, so retry send and receive,
-			 * to next nameserver on queue */
-			goto tryall;
+
+		ipass = 0;
+		ipos = 0;
+		while( ipos < ifds ) {
+			if( fds[ipos] == -1 || !FD_ISSET(fds[ipos], &fdsets) ) {
+				ipos++;
+				continue;
 		}
-#endif
 
-		len = recv(fd, packet, PACKETSZ, 0);
-		if (len < HFIXEDSZ) {
-			/* too short! */
-			goto again;
+			memset(packet, 0, PACKETSZ);
+			if( (len = recv(fds[ipos], packet, PACKETSZ, 0)) < HFIXEDSZ ) {
+				ipos++;
+				continue;
 		}
 
 		__decode_header(packet, &h);
@@ -941,131 +907,108 @@
 		DPRINTF("id = %d, qr = %d\n", h.id, h.qr);
 
 		if ((h.id != local_id) || (!h.qr)) {
-			/* unsolicited */
-			goto again;
+				ipos++;
+				continue;
 		}
 
-		DPRINTF("Got response (i think)!\n");
+		DPRINTF("Got response %s\n", "(i think)!");
 		DPRINTF("qrcount=%d,ancount=%d,nscount=%d,arcount=%d\n",
 				h.qdcount, h.ancount, h.nscount, h.arcount);
 		DPRINTF("opcode=%d,aa=%d,tc=%d,rd=%d,ra=%d,rcode=%d\n",
 				h.opcode, h.aa, h.tc, h.rd, h.ra, h.rcode);
 
-		if (h.rcode == NXDOMAIN) {
-// bug 660 says we treat negative response as an error and retry
-// which is, eh, an error. :) We were incurring long delays because of this.
-			/* if possible, try next search domain */
-			if (!ends_with_dot) {
-				int sdomains;
-				__UCLIBC_MUTEX_LOCK(__resolv_lock);
-				sdomains = __searchdomains;
-				__UCLIBC_MUTEX_UNLOCK(__resolv_lock);
-				if (variant < sdomains - 1) {
-					/* next search */
-					variant++;
+		if ((h.rcode) || (h.ancount < 1)) {
+				ipos++;
 					continue;
 				}
-			}
-			/* this is not an error - don't goto again! */
-			h_errno = HOST_NOT_FOUND;
-			goto fail1;
-		}
-		if (h.rcode != 0) /* error */
-			goto again;
-		/* code below won't work correctly with h.ancount == 0, so... */
-		if (h.ancount < 1) {
-			h_errno = NO_DATA; /* is this correct code? */
-			goto fail1;
-		}
 
 		pos = HFIXEDSZ;
-
+			i = -1;
 		for (j = 0; j < h.qdcount; j++) {
 			DPRINTF("Skipping question %d at %d\n", j, pos);
-			/* returns -1 only if packet == NULL (can't happen) */
 			i = __length_question(packet, pos);
-			//if (i < 0)
-			//	goto again;
 			DPRINTF("Length of question %d is %d\n", j, i);
-			pos += i;
+				if( i < 0 ) break;
+				else pos += i;
+		}
+			
+			if( i < 0 ) {
+				ipos++;
+				continue;
 		}
+			
 		DPRINTF("Decoding answer at pos %d\n", pos);
 
 		first_answer = 1;
-		for (j = 0; j < h.ancount && pos < len; j++) {
-			i = __decode_answer(packet, pos, len, &ma);
-			if (i < 0) {
+			for( j = 0; j < h.ancount; j++, pos += i ) {
+				if( (i = __decode_answer(packet, pos, len, &ma)) < 0 ) {
 				DPRINTF("failed decode %d\n", i);
-				/* if the message was truncated and we have
-				   decoded some answers, pretend it's OK */
-				if (j && h.tc)
 					break;
-				goto again;
 			}
-			pos += i;
 
 			if (first_answer) {
 				ma.buf = a->buf;
 				ma.buflen = a->buflen;
 				ma.add_count = a->add_count;
 				memcpy(a, &ma, sizeof(ma));
-				if (a->atype != T_SIG && (NULL == a->buf || (type != T_A && type != T_AAAA)))
-					break;
-				if (a->atype != type)
+					if( a->atype != T_SIG && (0 == a->buf || (type != T_A && type != T_AAAA)) ) break;
+					if( a->atype != type ) {
 					continue;
+							}
 				a->add_count = h.ancount - j - 1;
-				if ((a->rdlength + sizeof(struct in_addr*)) * a->add_count > a->buflen)
-					break;
+					
+					if( (a->rdlength + sizeof(struct in_addr*)) * a->add_count > a->buflen ) break;
+					
 				a->add_count = 0;
 				first_answer = 0;
 			} else {
 				free(ma.dotted);
-				if (ma.atype != type)
-					continue;
+					if( ma.atype != type ) continue;
 				if (a->rdlength != ma.rdlength) {
 					free(a->dotted);
 					DPRINTF("Answer address len(%u) differs from original(%u)\n",
 							ma.rdlength, a->rdlength);
-					goto again;
+						i = -1;
+						break;
 				}
 				memcpy(a->buf + (a->add_count * ma.rdlength), ma.rdata, ma.rdlength);
 				++a->add_count;
 			}
 		}
 
+			if( i < 0 ) {
+				ipos++;
+				continue;
+			}
+
 		DPRINTF("Answer name = |%s|\n", a->dotted);
 		DPRINTF("Answer type = |%d|\n", a->atype);
 
-		close(fd);
+			ipass = 1;
+			break;
+		}
+
+		if( ipass ) {
+			for( ipos = 0; ipos < ifds; ipos++ )
+				if( fds[ipos] != -1 ) {
+					close(fds[ipos]);
+					fds[ipos] = -1;
+				}
+
+			if( outpacket ) *outpacket = packet;
+			else free(packet);
 
-		if (outpacket)
-			*outpacket = packet;
-		else
-			free(packet);
 		free(lookup);
 
 		/* Mess with globals while under lock */
 		__UCLIBC_MUTEX_LOCK(mylock);
-		static_ns = local_ns;
-		static_id = local_id;
+		id = local_id;
 		__UCLIBC_MUTEX_UNLOCK(mylock);
 
-		return len;				/* success! */
-
- tryall:
-		/* if there are other nameservers, give them a go,
-		   otherwise return with error */
-		variant = -1;
-		local_ns = (local_ns + 1) % nscount;
-		if (local_ns == 0)
-			retries++;
-
-		continue;
+		return (len);				/* success! */
+		}
 
- again:
 		/* if there are searchdomains, try them or fallback as passed */
-		if (!ends_with_dot) {
-			int sdomains;
 			__UCLIBC_MUTEX_LOCK(__resolv_lock);
 			sdomains = __searchdomains;
 			__UCLIBC_MUTEX_UNLOCK(__resolv_lock);
@@ -1072,30 +1016,28 @@
 			if (variant < sdomains - 1) {
 				/* next search */
 				variant++;
-				continue;
-			}
-		}
-		/* next server, first search */
-		local_ns = (local_ns + 1) % nscount;
-		if (local_ns == 0)
-			retries++;
+		    } else {
+			/* next servers, first search */
 		variant = -1;
+		   --itrys;
+		}
 	}
 
- fail:
-	h_errno = NETDB_INTERNAL;
- fail1:
-	if (fd != -1)
-		close(fd);
-	free(lookup);
-	free(packet);
-	/* Mess with globals while under lock */
-	if (local_ns != -1) {
 		__UCLIBC_MUTEX_LOCK(mylock);
-		static_ns = local_ns;
-		static_id = local_id;
+	    id = local_id;
 		__UCLIBC_MUTEX_UNLOCK(mylock);
+
+ fail:
+	for( ipos = 0; ipos < ifds; ipos++ )
+		if( fds[ipos] != -1 ) {
+			close(fds[ipos]);
+			fds[ipos] = -1;
 	}
+		
+	if( lookup ) free(lookup);
+	if( packet ) free(packet);
+	h_errno = NETDB_INTERNAL;
+
 	return -1;
 }
 #endif
