diff -ruN ppp-2.4.4.orig/pppd/auth.c ppp-2.4.4/pppd/auth.c
--- ppp-2.4.4.orig/pppd/auth.c	2009-10-13 22:22:41.000000000 +0400
+++ ppp-2.4.4/pppd/auth.c	2009-10-16 22:34:34.000000000 +0400
@@ -2304,9 +2304,11 @@
 
     if (fstat(fileno(f), &sbuf) < 0) {
 	warn("cannot stat secret file %s: %m", filename);
+#if 0
     } else if ((sbuf.st_mode & (S_IRWXG | S_IRWXO)) != 0) {
 	warn("Warning - secret file %s has world and/or group access",
 	     filename);
+#endif
     }
 }
 
