Index: microdc2-0.15.6-20121011/lib/quotearg.c
===================================================================
--- microdc2-0.15.6-20121011.orig/lib/quotearg.c	2012-11-11 15:24:10.000000000 +0400
+++ microdc2-0.15.6-20121011/lib/quotearg.c	2012-11-11 15:30:26.755376089 +0400
@@ -495,6 +495,25 @@
 
             bool printable;
 
+            /* Dirty hack - zyxmon. Test if UTF-8 2 or 3 bytes char */
+            if (((c & 0xe0) == 0xc0) && ((arg[i+1] & 0xc0) == 0x80))
+            {
+                STORE(c);
+                c=arg[i+1];
+                i++;
+                goto store_c;
+            }
+            if (((c & 0xf0) == 0xe0) && ((arg[i+1] & 0xc0) == 0x80) && ((arg[i+2] & 0xc0) == 0x80))
+            {
+                STORE(c);
+                c=arg[i+1];
+                i++;
+                STORE(c);
+                c=arg[i+1];
+                i++;
+                goto store_c;
+            }
+
             if (unibyte_locale)
             {
                 m = 1;
