diff -ruN dnrd-2.20.3.orig/src/udp.c dnrd-2.20.3/src/udp.c
--- dnrd-2.20.3.orig/src/udp.c	2010-03-13 01:41:42.000000000 +0300
+++ dnrd-2.20.3/src/udp.c	2010-03-26 16:44:33.000000000 +0300
@@ -59,8 +59,8 @@
 		sizeof(struct sockaddr_in));
 
     if (rc != len) {
-	log_msg(LOG_ERR, "sendto error: %s: ",
-		inet_ntoa(srv->addr.sin_addr), strerror(errno));
+	//log_msg(LOG_ERR, "sendto error: %s: ",
+	//	inet_ntoa(srv->addr.sin_addr), strerror(errno));
 	return (rc);
     }
     if ((srv->send_time == 0)) srv->send_time = now;
