--- a/ui/xupnpd_ui.lua
+++ b/ui/xupnpd_ui.lua
@@ -28,7 +28,7 @@ function ui_handler(args,data,ip)
                     local fname=string.match(string.sub(tt[1],1,n-1),'filename=\"(.+)\"')
 
                     if fname and string.find(fname,'.+%.m3u$') then
-                        local tfname='/tmp/'..fname
+                        local tfname='/media/DISK_A1/system/tmp/'..fname
 
                         local fd=io.open(tfname,'w+')
                         if fd then
